### PR TITLE
Add Enemy Elemental Resistance (old PR#238)

### PR DESCRIPTION
**Pull request by h-yasha**
Recreate Branch, as I screwed up.
old PR: #238.

### CODE REVIEW

**Comment by kei-gbf** [comment]
@h-yasha 

- [ ] I saw before, this was implemented to damageUp variable. but now not.
    How does it work for supplemental damage now ?
- [x] debuf info show Enemy Elemental Resistance
- [ ] not implement show the info per character ? (just confirm)
- [x] case Element "Non" and "Non (but Crit)"
- [ ] check real enemy's Enemy Elemental Resistance if anyone have the list
  - [ ] step 5% is ok?

works fine,
I just want to know about how that work to supplemental damage,
and how to check it correct formula.
other are just note extra topics (not change request)

**Comment by h-yasha** [comment]
https://github.com/MotocalDevelopers/motocal/pull/238#issuecomment-499957701

will test how it interacts with supplemental damage. _(unfortunately Extreme Lignoid, trial battle does not have elemental resistance as it "supposedly" should have. oh well, waste shiva hosts it is)_

**Comment by h-yasha** [comment]
- [ ] Correct the interaction with supplemental damage.

**Comment by kei-gbf** [comment]
- [ ] missing remove the last character
```json
"en": "Supplemental Damage Source)"
```

- [ ] code clean up, this is prettier show, or any style you like.
  improve point is easy to find where Math.max close,
  to ensure it's not missing damageUp/enemyResistance for total[key].
```js
var supplementalDamageBuff = Math.ceil(
    Math.max(
        totals[key]["supplementalDamageBuff"],
        buff["supplementalDamageBuff"]
    )
        * (1 + damageUP) * (1 - enemyResistance)
);
```

update supplemental damage logic ... ok
- [x] supplementalDamageBuff
  - [x] totals[key] ... just note, code can be clean up, i needed to check where Math.max close ")"
  - [x] buff
- [x] supplementalThirdHit source .... support for Hallesena
- convent
  - [x] impervious/不壊の誓約
  - [x] victorious/凱歌の誓約
  - [x] contentious/修羅の誓約
  - [x] deleterious/致命の誓約
  - [x] calamitous/災禍の誓約

**Comment by kei-gbf** [comment]
- [ ] check other missing enemyResistance

in module.exports.calcOugiDamage  missing enemyResistance for damageUp > 0 ?
```js
    // The final damage becomes the correction amount + the minimum attenuation line
    damage = damage + overedDamage;

    // Damage raised
    if (damageUP > 0) {
        return (1.0 + damageUP) * damage;
    }
    return damage * (1.0 - enemyResistance);
```

- minor change
  - supplemental damages:  (1 + damageUP) * (1 - enemyResistance)
  - could be 1.0 if damageUp and enemyResistance were float type.
    that can show type information of variable implicitly.
- remove if/return and  can change the order too
  - damageUp = Math.max(1, damageUp); // XXX: this can ignore negative case ?
  - return damage * (1.0 + damageUP) * (1.0 - enemyResistance);

Those changes make `damage * (1.0 + damageUP) * (1.0 - enemyResistance);` expression
search-able in text editor.

**Comment by kei-gbf** [comment]
additonalDamage does not contain damageUp?

----
- [x] if damageUp > 0 did not return in `module.exports.calcDamage`
```js
    var res = damage + overedDamage;

    if (additionalDamage > 0) {
        res *= 1.0 + additionalDamage
    }

    // "Granted Damage Increase" is a correction after attenuation
    if (damageUP > 0) {
        res *= 1.0 + damageUP
    }

    return res * (1.0 - enemyResistance);
```

**Comment by kei-gbf** [comment]
@h-yasha 

Can you check damage formula 
@kagekiyo7 reported, wiki(ja) formula 

calcOugiDamage should take damage which calculated after damage cap etc
but Motocal take summedAttack and re-calc defense etc in calcOugiDamage again.
I am not sure about wiki's info.

This may affect to supplemental damage too,
if the source damage had supplemental damage, then no need to add in Ougi, Chain burst 
(now need to add them, because it passed summedAttack which does not contain supplemental)

File: src/global_logic.js [hunk 1]
```diff
@@ -210,16 +210,12 @@ module.exports.calcDamage = function (summedAttack, totalSkillCoeff, criticalRat
...
-        res *= 1.0 + additionalDamage
-    }
-
+    res *= Math.max(1.0, 1.0 + additionalDamage);
     // "Granted Damage Increase" is a correction after attenuation
-    if (damageUP > 0) {
-        res *= 1.0 + damageUP
-    }
+    res *= Math.max(1.0, 1.0 + damageUP);
+    res *= Math.max(0.0, Math.min(1.0, 1.0 - enemyResistance));
```
Comment (kei-gbf) [inline src/global_logic.js]: nice, it's Diff friendly change :+1:

**Comment by kei-gbf** [comment]
I think it should be new PR to fix the base damage source of calcOugiDamage
Because that may change about supplemental damage code of ougi or chain burst.

current PR is ok as reviewed. it just found another issue.

**Comment by h-yasha** [comment]
Note: this is hard to implement currently, (would need be implemented first to account for this cases)
- in cases like Narumaya/Anila having Superior Element Additional Damage, or other cases where character has another element additional damage, such as Vira/Amira. the additional damage does not get reduced. in-game note does imply that too. it is attack's element based rather than character element.
image

**Comment by kei-gbf** [comment]
敵非有利耐性のレビュー完了です。

注意点: 属性追撃 等は現状では対応が難しい。

**Comment by kagekiyo7** [comment]
@h-yasha @kei-gbf 
Is it unnecessary to change this part? (add enemyResistance)
https://github.com/MotocalDevelopers/motocal/blob/1941c9ef105c65e18d8b90dbd4bb919b713a7e36/src/global_logic.js#L3203
https://github.com/MotocalDevelopers/motocal/blob/1941c9ef105c65e18d8b90dbd4bb919b713a7e36/src/global_logic.js#L3219

**Comment by kei-gbf** [comment]
@kagekiyo7 
thx ur assist


static code checker, unit tests can detect this kind of problem.
that we does not have now. currently it's hard to test every review time, by manual.

MEMO: when calc damage parameters are changed, need to change following:
- Haisui graph
- Simulation

**Comment by kei-gbf** [comment]
@h-yasha 

I was missing to check generateSimulationData 
supplemental damage also not implemented there?

**Pull request merged**